### PR TITLE
Lf 2803 irrigation post

### PR DESCRIPTION
**Pull request by wasswarichard**
https://lite-farm.atlassian.net/browse/LF-2803

### CODE REVIEW

**Comment by Duncan-Brain** [comment]
@wasswarichard 

Here's what I got so far without testing it...
1. taskController - looks like its going to interfere with @anant00729 LF-2818 maybe wait or coordinate with him after that is merged before this PR
2. TaskLocations/index - looks like you accidentally mixed this up from a different branch! This is from my most recent PR and doesnt belong here

File: packages/api/src/controllers/taskController.js [hunk 1]
```diff
@@ -245,14 +247,42 @@ const taskController = {
     }
   },
 
+  async checkCustomDependencies(typeOfTask, data, farm_id) {
```
Comment (JimmyRowland) [inline packages/api/src/controllers/taskController.js:250]: use withGraphFetched
https://vincit.github.io/objection.js/api/query-builder/eager-methods.html

Comment (JimmyRowland) [inline packages/api/src/controllers/taskController.js:250]: https://vincit.github.io/objection.js/api/query-builder/mutate-methods.html#upsertgraph
Documentation for upsert graph. These are already implemented by objection js. Just add proper relation mappings to task models and validations

**Comment by wasswarichard** [comment]
@anant00729 please look at this

File: packages/api/src/controllers/taskController.js [hunk 2]
```diff
@@ -295,11 +297,42 @@ const taskController = {
...
     switch (typeOfTask) {
       case 'field_work_task': {
         return await this.checkAndAddCustomFieldWork(data, farm_id);
       }
+      case 'irrigation_task':
+        if (data.irrigation_type?.irrigation_task_type_other) {
+          await IrrigationTypesModel.insertCustomIrrigationType({ ...irrigationTypeValues });
+        }
+        if (data.irrigation_type?.set_default_irrigation_task_type_measurement) {
+          await IrrigationTypesModel.query()
```
Comment (anant00729) [inline packages/api/src/controllers/taskController.js:315]: can shift this line to IrrigationTypesModel file. can add this in the static function.

File: packages/api/src/controllers/taskController.js [hunk 3]
```diff
@@ -295,11 +297,42 @@ const taskController = {
...
+            .where('irrigation_type_name', irrigationTypeValues.irrigation_type_name)
+            .patch({
+              default_measuring_type: irrigationTypeValues.default_measuring_type,
+              farm_id: irrigationTypeValues.farm_id,
+            })
+            .returning('*');
+        }
+        if (data.location_defaults) {
+          for (const location_default of data.location_defaults) {
+            await locationDefaultsModel.transaction(async (trx) => {
```
Comment (anant00729) [inline packages/api/src/controllers/taskController.js:325]: can shift this line to LocationDefaultsModel file. can add this in the static function.

File: packages/api/src/models/irrigationTypesModel.js [hunk 4]
```diff
@@ -21,5 +22,21 @@ class IrrigationTypesModel extends Model {
       additionalProperties: false,
     };
   }
+
+  static async insertCustomIrrigationType(row) {
+    // insert into irrigation_type table
+    await knex('irrigation_type').insert(row);
+
+    // update irrigation_task type column enum constraint
+    const irrigation_type_names = await knex('irrigation_type').select('irrigation_type_name');
```
Comment (anant00729) [inline packages/api/src/models/irrigationTypesModel.js:31]: use camel case e.g IrrigationTypeNames

**Review by anant00729 — CHANGES_REQUESTED** [review]

**Comment by anant00729** [comment]
Looks good to me

**Review by anant00729 — APPROVED** [review]

**Pull request merged**